### ISSUE TITLE
Ivy-ngcc: Fail to compile in rc.02

### ISSUE THREAD

**Issue by Aw3same** [angular/angular#30299]
# 🐞 bug report

### Is this a regression?
NO

### Description
When I'm trying to enable the Ivy compiler options, in the `postinstall `script, I receive an error:
![image](https://user-images.githubusercontent.com/10500082/57306333-ed707200-70e2-11e9-80b8-70cae06bd219.png)

I think it's related by [`ngx-markdown`](https://github.com/jfcere/ngx-markdown) library, but I cannot recognize the error. [Here](https://unpkg.com/ngx-markdown@7.1.5/fesm5/ngx-markdown.js) is the file I think it fails.

I've taken a look to this issue [#29835](https://github.com/angular/angular/issues/29835) but all seems to be ok.

## 🔬 Minimal Reproduction

```
ng new repro-app --skip-install 
cd repro-app 
```

Replace package.json with this [file](https://github.com/Aw3same/repro-err/blob/master/package.json)

`npm i `

## 🔥 Exception or Error
<pre><code>
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Users\\Aw3sam3\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'postinstall'
1 verbose cli ]
2 info using npm@6.9.0
3 info using node@v12.1.0
4 verbose run-script [ 'postinstall' ]
5 info lifecycle application@1.10.0~postinstall: application@1.10.0
6 verbose lifecycle application@1.10.0~postinstall: unsafe-perm in lifecycle true
7 verbose lifecycle application@1.10.0~postinstall: PATH: C:\Users\Aw3sam3\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin;C:\Users\Aw3sam3\Desktop\repro-app\node_modules\.bin;C:\Users\Aw3sam3\cmder\bin;C:\Users\Aw3sam3\cmder\vendor\bin;C:\Users\Aw3sam3\cmder\vendor\conemu-maximus5\ConEmu\Scripts;C:\Users\Aw3sam3\cmder\vendor\conemu-maximus5;C:\Users\Aw3sam3\cmder\vendor\conemu-maximus5\ConEmu;C:\Program Files\Python37\Scripts\;C:\Program Files\Python37\;c:\oracle\product\12.1.0\client_1;c:\oracle\product\12.1.0\client_1\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;c:\oracle64\product\12.1.0\client_1;c:\oracle64\product\12.1.0\client_1\bin;C:\ProgramData\Oracle\Java\javapath;c:\Program Files (x86)\Intel\iCLS Client\;c:\Program Files\Intel\iCLS Client\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\Intel\Intel(R) Management Engine Components\IPT;C:\Program Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files (x86)\Intel\Intel(R) Management Engine Components\IPT;C:\WINDOWS\System32\OpenSSH\;C:\Users\Aw3sam3\.dnx\bin;C:\Program Files\Microsoft DNX\Dnvm\;C:\Program Files (x86)\Windows Kits\8.1\Windows Performance Toolkit\;C:\Program Files\Microsoft SQL Server\120\Tools\Binn\;C:\Program Files\Git\cmd;C:\Program Files\Microsoft VS Code\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\nodejs\;C:\Program Files\Python37;C:\Python27;C:\Program Files\Microsoft\Web Platform Installer\;C:\Users\Aw3sam3\AppData\Local\Microsoft\WindowsApps;C:\Users\Aw3sam3\AppData\Roaming\npm;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Users\Aw3sam3\cmder
8 verbose lifecycle application@1.10.0~postinstall: CWD: C:\Users\Aw3sam3\Desktop\repro-app
9 silly lifecycle application@1.10.0~postinstall: Args: [ '/d /s /c', 'ivy-ngcc' ]
10 silly lifecycle application@1.10.0~postinstall: Returned: code: 1  signal: null
11 info lifecycle application@1.10.0~postinstall: Failed to exec postinstall script
12 verbose stack Error: application@1.10.0 postinstall: `ivy-ngcc`
12 verbose stack Exit status 1
12 verbose stack     at EventEmitter.<anonymous> (C:\Users\Aw3sam3\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
12 verbose stack     at EventEmitter.emit (events.js:196:13)
12 verbose stack     at ChildProcess.<anonymous> (C:\Users\Aw3sam3\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
12 verbose stack     at ChildProcess.emit (events.js:196:13)
12 verbose stack     at maybeClose (internal/child_process.js:1000:16)
12 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:267:5)
13 verbose pkgid application@1.10.0
14 verbose cwd C:\Users\Aw3sam3\Desktop\repro-app
15 verbose Windows_NT 10.0.17763
16 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\Aw3sam3\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "run" "postinstall"
17 verbose node v12.1.0
18 verbose npm  v6.9.0
19 error code ELIFECYCLE
20 error errno 1
21 error application@1.10.0 postinstall: `ivy-ngcc`
21 error Exit status 1
22 error Failed at the application@1.10.0 postinstall script.
22 error This is probably not a problem with npm. There is likely additional logging output above.
23 verbose exit [ 1, true ]
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/

Angular CLI: 8.0.0-rc.2
Node: 12.1.0
OS: win32 x64
Angular: 8.0.0-rc.2
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.0-rc.2
@angular-devkit/build-angular     0.800.0-rc.2
@angular-devkit/build-optimizer   0.800.0-rc.2
@angular-devkit/build-webpack     0.800.0-rc.2
@angular-devkit/core              8.0.0-rc.2
@angular-devkit/schematics        8.0.0-rc.2
@angular/cdk                      8.0.0-rc.0
@angular/material                 8.0.0-rc.0
@ngtools/webpack                  8.0.0-rc.2
@schematics/angular               8.0.0-rc.2
@schematics/update                0.800.0-rc.2
rxjs                              6.5.1
typescript                        3.4.5
webpack                           4.30.0
</code></pre>

**Comment by JoostK** [comment]
Thanks for reporting! You have been thorough in your investigation, your conclusion that `ngx-markdown` is a problem here is correct. It is indeed somewhat related to #29835, which was also because of an incompatibility due to ES5 downleveling.

For libraries that have been compiled with `downlevelIteration`, TypeScript does not generate `Array.prototype.slice()` calls (those are supported) but instead inserts a call to `tslib`'s `__spread` helper. We need to add basic support for statically interpreting those calls to resolve this issue. The problem with `tslib` mostly is that it could also be embedded into the file itself, in which case it is harder to identify and insert special behavior for it.

This is also tracked internally in FW-1298

**Comment by Aw3same** [comment]
Perfect, thank you very much for the quick response! I was already going crazy trying to find a solution. I suppose then that we simply have to wait. Is there any way to track the FW-1298 ?

Many thanks

**Comment by JoostK** [comment]
@Aw3same I marked this under FW-1298 and any upcoming PR will reference this issue number, so this issue should be closed automatically once a fix gets in :)

**Comment by SebastianPodgajny** [comment]
@JoostK I'm tried building my own lib with `downlevelIteration: false` and error still exist 🙁

**Comment by JoostK** [comment]
@SebastianPodgajny Maybe there's more reason to use `__spread`, I don't know for sure. Can you share the relevant snippets of code (basically NgModule metadata in the ES5 file) from your library, so we can get an idea of what the issue could be?

**Comment by SebastianPodgajny** [comment]
@JoostK 
this for example cause error
```
@NgModule({
  imports: [RtUICoreBaseModule, RtUICorePlatformModule],
  declarations: [...DIRECTIVES],
  exports: [...DIRECTIVES],
})
```

**Comment by SebastianPodgajny** [comment]
It happens on `fesm5 as esm5` step, one thing I found is that if I disable `importHelpers` this step will pass and error will happen on `esm5 as esm5` step so:
enabled `importHelpers` (default ng-packagr)
`fesm2015 as esm2015` ✅ 
`fesm5 as esm5` ❌ 
`esm2015 as esm2015`  ❌ 
`esm5 as esm5`  ❌ 

importHelpers: false
`fesm2015 as esm2015` ✅ 
`fesm5 as esm5` ✅ 
`esm2015 as esm2015`  ✅ 
`esm5 as esm5`  ❌

**Comment by Aw3same** [comment]
Maybe this error was reported, but I want to know why i'm receiving it, and if it's related by this issue:
![image](https://user-images.githubusercontent.com/10500082/57360928-99fe3280-717b-11e9-8ae5-2c64f936cdd7.png)

I have the changes in `tsconfig.app.json` :
```
	"angularCompilerOptions": {
		"enableIvy": true,
		"allowEmptyCodegenFiles": true
	},
```
and my `app.module` look like this:

```
import { registerLocaleData } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import localeEs from '@angular/common/locales/es';
import { APP_INITIALIZER, LOCALE_ID, NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterModule } from '@angular/router';
import { LeafletModule } from '@asymmetrik/ngx-leaflet';
import { LeafletDrawModule } from '@asymmetrik/ngx-leaflet-draw';
import { ToastrModule } from 'ngx-toastr';
import { CoreModule } from './@core/core.module';
import { SharedModule } from './@shared-layout/shared.module';
import { App } from './app.component';
import { routing } from './app.routing';
import { MyAppLoadService } from './my-app-load.service';
import { PagesModule } from './pages/pages.module';

export function init_app(firstLoadService: MyAppLoadService) {
	return () => firstLoadService.initializeMyApp();
}

registerLocaleData(localeEs, 'es');

// Application wide providers
const APP_PROVIDERS = [
	{
		provide: LOCALE_ID,
		useValue: 'es-ES',
	},
	MyAppLoadService,
	{
		provide: APP_INITIALIZER,
		useFactory: init_app,
		deps: [MyAppLoadService],
		multi: true,
	},
];

/**
 * `AppModule` is the main entry point into Angular's bootstraping process
 */
@NgModule({
	bootstrap: [App],
	declarations: [App],
	imports: [
		// import Angular's modules
		BrowserModule,
		BrowserAnimationsModule,
		HttpClientModule,
		RouterModule,
		FormsModule,
		ReactiveFormsModule,
		routing,
		SharedModule,
		PagesModule,
		CoreModule,
		LeafletModule.forRoot(),
		LeafletDrawModule.forRoot(),
		ToastrModule.forRoot({
			timeOut: 20000,
			closeButton: true,
			progressBar: true,
			preventDuplicates: true,
			progressAnimation: 'decreasing',
			positionClass: 'toast-bottom-left',
			tapToDismiss: true,
			enableHtml: true,
		}),
	],
	providers: [APP_PROVIDERS],
})
export class AppModule {
	constructor() {}
}

```

**Comment by JoostK** [comment]
@Aw3same I can't tell from that screenshot/code sample, is there a (minimal) reproduction available?

**Comment by adrogon** [comment]
@Aw3same `ngx-markdown` has provided a fix in `8.0.2-beta.0` (see https://github.com/jfcere/ngx-markdown/issues/159).
The commit fixing it being https://github.com/jfcere/ngx-markdown/pull/160/commits/3a056bc7d8dca5bd730b4f0b68404f69e6a3f1d9 which is about... removing the spread operator.

It's a good workaround, but prohibiting the use of a common operator does not sound like a good solution. I hope there's a proper resolution for this in the near future.

Also, https://github.com/angular/angular/pull/30771 makes the error message you received more self-explanatory.

**Comment by JoostK** [comment]
@adrogon the relevant PR is here: #30492, we hope to get it in soon. And yes, FW-1298 is only private but the Github PR is where the activity happens anyway.
